r1cs/gadgets/std/src/bits/boolean.rs

```diff
@@ -437,25 +437,29 @@ impl Boolean {
     {
         let field_elements: Vec<ConstraintF> =
             ToConstraintField::<ConstraintF>::to_field_elements(values).unwrap();
-        let values_len = values.len();
-
+        let modulus_size = ConstraintF::size_in_bits();
         let max_size = ConstraintF::Params::CAPACITY as usize;
+
         let mut allocated_bits = Vec::new();
-        for (i, field_element) in field_elements.into_iter().enumerate() {
+
+        for (i, (field_element, bit_chunk)) in field_elements
+            .into_iter()
+            .zip(values.chunks(max_size))
+            .enumerate()
+        {
             let fe = FpGadget::<ConstraintF>::alloc_input(&mut cs.ns(|| format!("Field element {}", i)), || {
                 Ok(field_element)
             })?;
+
             let fe_bits = fe.to_bits(cs.ns(|| format!("Convert fe to bits {}", i)))?;
```

Comment:
Use variant with length restricted to avoid accepting any value for pub inputs playing with the padding bits. Do the same with Uint8::alloc_input_vec